components/portal/widgets/controllers.js

```diff
@@ -414,41 +414,43 @@ define(['angular'], function(angular) {
 
       var now = new Date();
       var templateIndex = indexWithinTemplateRange(now, callsToAction);
-      if (-1 < templateIndex) {
+      if (templateIndex > -1) {
         displayTimeSensitiveContent(callsToAction[templateIndex]);
       }
     };
 
     var resolveDate = function(str, defaultEndOfDay) {
-      var result = null;
-      if (str) {
-        var dateStr =
-          $filter('filterForDateWithYear')(str);
-        if (dateStr) {
-          if (dateStr.indexOf('T') < 0) {
-            if (defaultEndOfDay) {
-              dateStr += 'T23:59:59';
-            } else {
-              dateStr += 'T00:00:00';
-            }
-          }
-          result = new Date(dateStr);
+      // if it is valid ISO 8601 pass through
+      var fullISODate = moment(str, moment.ISO_8601, true);
+      if (fullISODate.isValid()) {
+        return fullISODate.toDate();
+      }
+
+      // if it is a partial ISO string, fill in missing fields
+      var partialISODate = moment(str, ['YYYY-MM-DD', 'MM-DD'], true);
+      if (partialISODate.isValid()) {
+        if (defaultEndOfDay) {
+          return partialISODate.endOf('day').toDate();
+        } else {
+          return partialISODate.startOf('day').toDate();
```

Comment:
please don't have multiple returns in a function. define the base case, then specify the outliers.

Reply:
I feel like we've had this discussion before.
Single return makes a lot of sense when dealing with side effects.
multiple returns or early return reduces complexity when dealing with pure functions.
This is a pure function, multiple returns makes sense.

Reply:
yes, we have, but I'm having a difficult time seeing the practical advantage. Was the function not "pure" with a single return statement? There were no side effects, the only mutations were scoped within the function. The two implementations are almost the same in terms of complexity: http://jsmeter.info/cs6kwf/1

With multiple return statements, the code is less malleable and bugs take longer to diagnose. The bug that precipitated this refactor was itself a result of having multiple `showBasicWidget()` paths, one `displayTimeSensitiveContent()` path, and a not immediately noticeable path where neither of them were run. To see that, I had to trace through the code to find that case, and not rely on the assumption that the code was hitting one of those prior cases.  Your function ends with a return of an invalid date, which would be an acceptable default, if any of the other code actually checked `isValid`. That aside, how quickly would you notice if that last return happened to be missing? With three other returns in the function, it wouldn't be immediately noticeable. If there's only one return statement, you know immediately if it's incorrect.

I'm not bringing this up to be contrarian. I'm trying to optimize for resiliency and ease of maintenance, and as of yet I don't see how multiple return statements do that.

Reply:
> Was the function not "pure" with a single return statement?

it is pure either way.

> With multiple return statements, the code is less malleable and bugs take longer to diagnose. 

I agree that quick diagnostic and malleability are good.
Multiple returns are more malleable. Single return tends to cause deeply nested conditions that are a nightmare to debug. Multiple returns encourages flat code, with conditions at the top level.

> bugs take longer to diagnose

How? Either way the first thing a developer should do is add a breakpoint and step through the code.
With multiple returns there is less complexity to debug.

> if any of the other code actually checked `isValid`

Why check `isValid`?
`moment` handles invalid automatically.
https://momentjs.com/docs/#invalid-moments

> at aside, how quickly would you notice if that last return happened to be missing?

Pretty quickly, Chrome dev tools highlight what ever value in a scope is.
Step through would catch this pretty quick.

> If there's only one return statement, you know immediately if it's incorrect.

Value is `underfined` there wasn't a `return`, multiple returns debugging seems pretty instant to me.

> I'm trying to optimize for resiliency and ease of maintenance

Same. Again agree philosophically, but differ in implementation.

> and as of yet I don't see how multiple return statements do that.

I've used single `return` in the past.
Deep conditionals that are hellish to debug tend to build up overtime.
Keeping flat conditionals using multiple `return`s and `throw`s avoids that.